### ISSUE TITLE
🔴 xformers not found! Please install it before trying to use it.

### ISSUE THREAD

**Issue by IronWolve** [oobabooga/text-generation-webui#1252]
### Describe the bug

Did both pip install xformers, and xformers in /text-generation-webui directory.
Evan ran the installer again.

https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Xformers

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

xformers missing

### Logs

```shell
🔴 xformers not found! Please install it before trying to use it.
Loading the extension "api"... Ok.
Loading the extension "sd_api_pictures"... Ok.
Loading the extension "gallery"... Ok.
Loading the extension "send_pictures"... Starting KoboldAI compatible api at http://127.0.0.1:5000/api
Ok.
```

### System Info

```shell
AMD 5950x, Nvidia 4090, Win11
```

**Comment by Ph0rk0z** [comment]
yes but did you install it inside conda or the correct venv?

**Comment by Ph0rk0z** [comment]
maybe this check fails on windows

```
try:
        import xformers.ops
    except Exception:
        print("🔴 xformers not found! Please install it before trying to use it.", file=sys.stderr)
```

**Comment by Gitbreast** [comment]
Same situation, why?

**Comment by Ph0rk0z** [comment]
Open up a python .. write` import xformers.ops`, what does it say?

**Comment by Gitbreast** [comment]
![image](https://user-images.githubusercontent.com/96460796/232319328-aa4dddbf-1e8a-4b2d-8e97-7833afa606e5.png)
@Ph0rk0z I got this, means I need to upgrade something?

**Comment by Ph0rk0z** [comment]
Yea, you have CPU pytorch.

**Comment by Ketorami** [comment]
I have the same problem. My notifications are like this.
![Screenshot 2023-04-17 134612](https://user-images.githubusercontent.com/101268214/232405714-e0f718d2-c7e6-47e9-aa92-15d03a3b8fa3.png)
but when i open the WebUI , I still get error that not found

![Screenshot 2023-04-17 134856](https://user-images.githubusercontent.com/101268214/232406119-0e85015c-ce75-4d9b-a30d-458f4da67a19.png)

**Comment by Ph0rk0z** [comment]
Triton doesn't work on windows I think.

**Comment by Gitbreast** [comment]
> Yea, you have CPU pytorch.

So I need to install a PyTorch for GPU and upgrade Python?

**Comment by Ketorami** [comment]
You can install PyTorch as this repo 
https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/8691#discussion-4972715

**Comment by Ph0rk0z** [comment]
Basically.. you need a cuda pytorch to do GPU inference.

**Comment by EnviralDesign** [comment]
the xformers library def needs to be installed in the env, which is managed by mamba. I was able to get it working with these steps, hopefully it helps:

1) run micromamba-cmd.bat, this will open a terminal, in the text-generation-webui's env activated already.
2) run `pip install xformers`
3) close that terminal, and close, then restart webui `start-webui.bat`

**Comment by Ketorami** [comment]
> the xformers library def needs to be installed in the env, which is managed by mamba. I was able to get it working with these steps, hopefully it helps:
> 
> 1. run micromamba-cmd.bat, this will open a terminal, in the text-generation-webui's env activated already.
> 2. run `pip install xformers`
> 3. close that terminal, and close, then restart webui `start-webui.bat`

Can you elaborate? Where is "text-generation-webui's env"? in oobabooga-windows\installer_files? And how i run micromamba-cmd.bat?

**Comment by EnviralDesign** [comment]
The env and the mamba stuff is here:
![image](https://user-images.githubusercontent.com/10091486/232841088-8d493a59-8639-4d29-98e6-e33daa35d081.png)

I'm on windows, so for me the micromamba-cmd.bat is located here. I was able to double click and it opened up ready to use.
![image](https://user-images.githubusercontent.com/10091486/232840917-ae0d9fb0-dd9c-43b6-9778-eb549e300f1a.png)

**Comment by Ketorami** [comment]
> สิ่ง env และ mamba อยู่ที่นี่: ![ภาพ](https://user-images.githubusercontent.com/10091486/232841088-8d493a59-8639-4d29-98e6-e33daa35d081.png)
> 
> ฉันใช้ windows ดังนั้นสำหรับฉัน micromamba-cmd.bat จะอยู่ที่นี่ ฉันสามารถคลิกสองครั้งและมันก็เปิดขึ้นพร้อมใช้งาน ![ภาพ](https://user-images.githubusercontent.com/10091486/232840917-ae0d9fb0-dd9c-43b6-9778-eb549e300f1a.png)

![Screenshot 2023-04-18 232435](https://user-images.githubusercontent.com/101268214/232842135-b909a6b4-17ae-4575-9873-3e887a157b1e.png)

What?

**Comment by Ketorami** [comment]
![Screenshot 2023-04-18 233412](https://user-images.githubusercontent.com/101268214/232844537-12eb9303-bc14-4b4b-8705-5545cbbc3702.png)
Sad TT

**Comment by EnviralDesign** [comment]
that's weird! I'm not sure. fyi for context I used the one click installer to set everything up, not sure if it would behave differently if you installed it another way or manually.

**Comment by tonyzehs** [comment]
I used the new installer zip, did a fresh install in a directory with no spaces. That worked for me.

**Comment by Jonseed** [comment]
After installing xformers, I get the Triton not available message, but it will still load a model and the webui. It also says "Replaced attention with xformers_attention" so it seems xformers is working, but it is not any faster in tokens/sec than without --xformers, so I don't think it is completely functional. What is Triton?

**Comment by tonyzehs** [comment]
> After installing xformers, I get the Triton not available message, but it will still load a model and the webui. It also says "Replaced attention with xformers_attention" so it seems xformers is working, but it is not any faster in tokens/sec than without --xformers, so I don't think it is completely functional. What is Triton?

Search Triton in the existing issues on github, I think someone answered it being a Linux specific thing. And yeah, xformers is working for you now.

**Comment by Jonseed** [comment]
If it is specific to Linux, why is it trying to run Triton on Windows? And if xformers is working, why is the inference speed not any faster than without it? I get about 3 tokens/s with or without it with my 3060 12GB card.

**Comment by tonyzehs** [comment]
> If it is specific to Linux, why is it trying to run Triton on Windows? And if xformers is working, why is the inference speed not any faster than without it? I get about 3 tokens/s with or without it with my 3060 12GB card.

Maybe the base code is the same across platforms, so updating/maintaining is easier for the devs? As for the speed increase, I personally found it to also be minimal, unlike Stable Diffusion. If anything, when running larger models the AI generations seem to spit out things that make less sense regarding context. here's a thread of people experimenting with it:

https://github.com/oobabooga/text-generation-webui/pull/950

**Comment by Ph0rk0z** [comment]
It was mentioned xformers may use less memory.

**Comment by Stoxis** [comment]
> The env and the mamba stuff is here:

I'm using the new installation, that batch script doesn't exist anymore. 
How would you go about downloading the library now? Thank you!

**Comment by bkutasi** [comment]
I added xfromers into `oobabooga_windows/text-generation-webui/requrments.txt` and it installed the latest xfromers, but I get a triton error when I use the --xformer parameter.
First 4 lines at launch:
`Gradio HTTP request redirected to localhost :)
A matching Triton is not available, some optimizations will not be enabled.
Error caught was: No module named 'triton'
bin D:\Other_projects\oobabooga\installer_files\env\lib\site-packages\bitsandbytes\libbitsandbytes_cuda117_nocublaslt.dll`
The speed doesn't change, I tried reinstalling it with the new release, but it's the same.

**Comment by barleyj21** [comment]
Yes, just today I tried to reinstall oobabooga with oobabooga_windows.zip and despite that I answered **_A_** to install for nvidia GPU, it still installed pytorch for cpu into venv. Used this https://github.com/oobabooga/text-generation-webui/issues/1252#issuecomment-1511211024 from this thread to reinstall it and then installed "pip3 install xformers" - it's now 0.0.19 version by default.
Now it's working, even though it's still giving "no triton" error

**Comment by MistakingManx** [comment]
> the xformers library def needs to be installed in the env, which is managed by mamba. I was able to get it working with these steps, hopefully it helps:
> 
> 1. run micromamba-cmd.bat, this will open a terminal, in the text-generation-webui's env activated already.
> 2. run `pip install xformers`
> 3. close that terminal, and close, then restart webui `start-webui.bat`

This broke my installation completely, because it was successful ;-;
Well I mean it still works, but it's significantly slower "triton" doesn't exist, and I don't think it's doing what it's supposed to.
